### DISCUSSION TITLE
RP2040 not recognized in run mode

### DISCUSSION THREAD

**Discussion by RyzerPhenix** [earlephilhower/arduino-pico#2565] [General]
I have made a custom PCB with a RP2040 on it.
I am able to program the RP2040 by holding the BOOTSEL button while plugging it in, then it shows up as a drive in Windows Explorer. I can upload my code and the microcontroller resets correctly and executes the code afterwards.
But, also while the RP2040 is not in boot mode, the Device does not show up in the device manager and I get an error there:£
"This device has stopped because it reported errors. (Code 43)
USB device descriptor request failed."
The USB Cable works fine, and everything seems to work, since I am able to upload the code.
Any help is much appreciated.

**Comment by earlephilhower** [comment]
The app on the Pico is responsible for running the USB connection, so if the app crashes you will lose USB.

I'd suggest starting out with one of the simple Arduino examples. If that doesn't give a good USB connection, your board may have some issues with the chip and its connections.

If that works, but your app doesn't, then you're going to need to debug the app.  SWD and a DebugProbe (nee PicoProbe) are useful there.

**Reply by RyzerPhenix** [reply]
I have allrwady tried with just Serial.begin() and Serial.println("Hello World!") in the loop and this doesn't work either. I can exclude my USB Connection, since I am able to upload to the RP2040.
Is there any Firmware uploaded eventually? If yes, am I able to upload just the firmware.
Also I read in the docs of the SDK, that you only have to press the BOOTSEL on the first upload, but mine doesn't even shiw up as a seriel port, I just get the Windows Error every time.
I am also able to provide the schematic and pictures of my board and the Arduino Code if that helps.

**Reply by maxgerhardt** [reply]
Can you show the schematic? What flash chip do you use? If you have a different flash chip than the Pico but still select the Pico as your board, then the Boot2.S file might be wrong and incompatible with your flash chip. The bootrom of the RP2040 works fine, so at least the most generic flash support code works good enough for uploading.

**Comment by RyzerPhenix** [comment]
[Schematic BüPX - OdinsRage Motherboard V2.0.pdf](https://github.com/user-attachments/files/17577427/Schematic.BuPX.-.OdinsRage.Motherboard.V2.0.pdf)

I just coppied the original Pi Pico Schematic, I used the exact same parts, so there is no possible issue. In the Schematic you will find that there are two RP2040s, for now, just concentrate on uC1. Also there is a USB Hub Chip connected to the USB-C Connector, I replaced that with a mechanical switch for now, since I didn't get the Hub Chip to work. The Cables to the Switch are isolated, so there is also no issue.

**Reply by maxgerhardt** [reply]
Can you provide a screenshot of the Arduino IDE's "Tools" menu so we can see how you configured your build?
And the flash chips on the PCB are 100% a W25Q16JVUXIQ?

**Reply by RyzerPhenix** [reply]
Yes. the Chip is exactly the same as the Picos

**Reply by RyzerPhenix** [reply]
![image](https://github.com/user-attachments/assets/3ef558e4-739c-4128-91e4-bc13c773b3d6)
Here my tools section.

**Reply by maxgerhardt** [reply]
I see. Well, the schematic aren't 100% equal to a Pico, e.g., [on a Pico](https://datasheets.raspberrypi.com/pico/pico-datasheet.pdf), the AVDD voltage divider uses 200 Ohm and 1 Ohm resistors, but you use both 200 Ohm resistors. But I guess that's not the real issue here. 

Let's strip out some variables and sources of error. Can you please upload theses UF2 files and see if either of them can get the LED blinking? It does not initialize the USB stack at all (`-DPIO_FRAMEWORK_ARDUINO_NO_USB`), but if successfull, it should blink the GPIO25 (your LED). One is for a generic flash chip, the other one is for what's used on the Pico.

[test_fws.zip](https://github.com/user-attachments/files/17577689/test_fws.zip)

Also, please try to answer within the created thread, not create a new one for each reply. The number of open threads in the discussion are otherwise exploding.

**Reply by earlephilhower** [reply]
Change to the `Generic rp2040` board and then select `Boot2`->`Generic SPI/4` to have the slowest, 1-bit flash interface possible on the chip.  If that doesn't work, then as @maxgerhardt suggested, your flash chip may not be wired or working properly.

You can use `picotool save -a -t bin out.bin` when in UF2 bootloader mode to dump what the Pico is "reading" out of flash.  (i.e. if flash isn't working you'll get junk or all 0/1s).  Looking at out.bin in a hex editor you should see the string `PicoOTA` and `otacommand.bin` in the 1st 12KB of data, and other C library strings later on.
````
...
000026b0: 9847 0500 fff7 82ff 029b 9847 019b 9847  .G.........G...G
000026c0: 3a00 2000 0399 009b 9847 a847 fff7 8eff  :. ......G.G....
000026d0: 05b0 f0bd 4946 0000 4558 0000 5250 0000  ....IF..EX..RP..
000026e0: 4643 0000 5069 636f 204f 5441 006f 7461  FC..Pico OTA.ota
000026f0: 636f 6d6d 616e 642e 6269 6e00 2f00 2e2e  command.bin./...
00002700: 006c 6974 746c 6566 7300 0000 ffff ffff  .littlefs.......
00002710: ffff ffff 0000 0000 6410 b71d c820 6e3b  ........d.... n;
00002720: ac30 d926 9041 dc76 f451 6b6b 5861 b24d  .0.&.A.v.QkkXa.M
00002730: 3c71 0550 2083 b8ed 4493 0ff0 e8a3 d6d6  <q.P ...D.......
...
````

**Reply by RyzerPhenix** [reply]
Sorry for opening new threads, I am new to Github Discussions _:()
To clarify, I am perfectly able to upload code to the RP2040, and the code is also running, but I can't get the Serial.println to output anything and windows gives me this error (is in german):
![image](https://github.com/user-attachments/assets/cdfc443b-9bca-4a17-b148-79669d8a123f)
When I check the Device Manager, I find this device:
![image](https://github.com/user-attachments/assets/9f228b7f-86c7-464c-bdc6-237e514e42da)
But there is no COM Port showing up. I allready tried updating all USB Drivers, but that didn't help...

**Reply by maxgerhardt** [reply]
Can you confirm that both `blink_no_usb_boot2_w25q080.uf2` and `blink_no_usb_boot2_generic_div_4.uf2` manage to blink the LED? (Ignore the USB port not opening)

**Reply by RyzerPhenix** [reply]
Yes both files work and let the LED blink.

**Reply by maxgerhardt** [reply]
I've never seen it that the UF2 Bootloader can connect via USB just fine, but code application can't (with the most simplest test code that you use). If the LED blinks fine with the no USB stack setting, then the flash is working, as well. Not sure what's going on there..

**Reply by RyzerPhenix** [reply]
That's the same thing I am braining about since weeks. The issue doesn't make any sense at all.
I think it could be a windows related issue, since the Device Manager says, that the USB-Device doesn't return a USB-Device descriptor...

**Reply by RyzerPhenix** [reply]
Okay, so I did get the uC2 to work. I don't know what I did, because I tried multiple times before, and it didn't work on BOTH uCs, I just tried my luck again and hit upload and now this uC is showing up as a Serial COM Port, I can access it via the Serial Monitor and it outputs what it should.
![Snapchat-609108307.jpg](https://github.com/user-attachments/assets/37645f6f-6ac1-4456-b66b-0a48decb8c21)

![Snapchat-1448635798.jpg](https://github.com/user-attachments/assets/1993dfde-2e14-447e-966d-641a10163096)

![Snapchat-582889920.jpg](https://github.com/user-attachments/assets/39e47041-78aa-4b2e-b382-a7eea0b43771)

Also there you can see my janky isolation of the USB lines to the switch to switch between uC1 and uC2. Propably that causes the issue, because now uC1 is just randomly connecting and disconnecting. I will try to make the USB lines more stable tommorow and report to this thread.

**Reply by RyzerPhenix** [reply]
I did get the RP2040 to fully work and send and receive data over the Serial Port. The issue in fact was the poor Isolation of the USB Wires I soldered onto the board instead of the USB Hub Chip. I soldered them closer to the board and hooked up the isolation to ground. That did the trick. Thanks for all help provided.

**Reply by maxgerhardt** [reply]
Interesting. I thought about whether there was a change in USB bitrate that made one USB device mode work but not the other but.. USB mass-storage-device (MSC, UF2 bootloader) and USB CDC (serial) should both be at 2.0 speeds, so 480 MBit/s, and not USB 1.1 bit rates of 12 MBit/s, as far as I know. I could have understood it a bit if the low-speed USB stuff works at bad isolation, btu the high-speed USB not. Hm. In any case, glad to see it being solved!

**Reply by earlephilhower** [reply]
> ...USB mass-storage-device (MSC, UF2 bootloader) and USB CDC (serial) should both be at 2.0 speeds...

On the RP2040 there's only a USB 1.1 PHY and controller so both should be 12Mb rate.  Anyway, if it's working we're all good, 1.1 or 2.0. :smile:

**Comment by RyzerPhenix** [comment]
[3D_PCB_BüPX - OdinsRage Motherboard V2.0.zip](https://github.com/user-attachments/files/17577459/3D_PCB_BuPX.-.OdinsRage.Motherboard.V2.0.zip)

I can also provide a 3D Model of the PCB itself.

**Comment by RyzerPhenix** [comment]
Here is the Code that is running:

```cpp
/*
* Ryzer OdinsRage Mouse with ADNS5050 Optical Mouse Sensor, Raspberry Pi RP2040 microncontroller and TI LP5036 RGB LED Controller
* This program is using the Mouse.h (Arduino), LP503X.h (RyzerPhenix [GitHub]) and adns5050.h (RyzerPhenix [GitHub]) libraries
* Autor: RyzerPhenix
* Date: 30.05.2024
* Version: 1.0
* ©2024 by Ryzer Industries
*/

#include "adns5050.h"
#include <FastLED.h>
#include <Mouse.h>

//Define Pins of the ADNS5050 Sensor
#define SDIO_PIN 12
#define SCLK_PIN 13
#define NCS_PIN 14
#define NRESET_PIN 15

//Define Mouse Buttons
#define RIGHT_BUTTON 10
#define LEFT_BUTTON 11
#define MIDDLE_BUTTON 9

//Define Mouse Wheel
#define ENCSIGA 2
#define ENCSIGB 3

//RGB
#define NUM_LEDS 6      /* The amount of pixels/leds you have */
#define DATA_PIN 4       /* The pin your data line is connected to */

//Create Objects
ADNS5050 MouseSensor(SDIO_PIN, SCLK_PIN, NCS_PIN, NRESET_PIN);
CRGBArray<NUM_LEDS> leds;

//X and Y Movement Variables
int deltaX, deltaY;
//Button State Variables
bool rightButton, leftButton, middleButton;
//Mouse Wheel Variables
int val;
int enc0Pos = 0;
int enc0ALast = LOW;
int n = LOW;

//Setup on the First Core
void setup() {
  Serial.begin(115200);
  stdio_usb_init();
  stdio_init_all();

  //Initialize Mouse
  Mouse.begin();

  //Initialize Sensor
  MouseSensor.begin();

  //Reset Sensor
  MouseSensor.reset();

  //Initialize Mouse Buttons
  pinMode(RIGHT_BUTTON, INPUT_PULLDOWN);
  pinMode(LEFT_BUTTON, INPUT_PULLDOWN);
  pinMode(MIDDLE_BUTTON, INPUT_PULLDOWN);
  
  //Initialize Rotary Encoder Pins
  pinMode(ENCSIGA, INPUT_PULLDOWN);
  pinMode(ENCSIGB, INPUT_PULLDOWN);
}

//Setup on the Second Core
void setup1() {
  //Initialize RGB LEDs
  FastLED.addLeds<NEOPIXEL, DATA_PIN>(leds, NUM_LEDS);
}

//Loop on the First Core
void loop() {
  //Functions for Operating the Mouse
  MouseMovement();
  MouseButtons();
  MouseWheel();
}

//Loop on the Second Core
void loop1() {
  CycleRGB();
}

//Mouse Movement Functions
void MouseMovement() {
  //Read the X- and Y-values of the DELTA_X (0x03) and DELTA_Y (0x04) Registers
  deltaX = MouseSensor.read(DELTA_X_REG);
  deltaY = MouseSensor.read(DELTA_Y_REG);

  //Mouse Movement
  if(deltaX != 0){
    Mouse.move(-deltaX, 0);
  }

  if(deltaY != 0){
     Mouse.move(0, deltaY);
  }
}

//Mouse Button Functions
void MouseButtons() {
  //Read Button States
  rightButton = digitalRead(RIGHT_BUTTON);
  leftButton = digitalRead(LEFT_BUTTON);
  middleButton = digitalRead(MIDDLE_BUTTON);

  //Button Presses
  //Right Button
  if(rightButton == HIGH){
    Mouse.press(MOUSE_RIGHT);
  }
  else if(rightButton == LOW){
    Mouse.release(MOUSE_RIGHT);
  }

  //Left Button
  if(leftButton == HIGH){
    Mouse.press(MOUSE_LEFT);
  }
  else if(leftButton == LOW){
    Mouse.release(MOUSE_LEFT);
  }

  //Middle Button
  if(middleButton == HIGH){
    Mouse.press(MOUSE_MIDDLE);
  }
  else if(middleButton == LOW){
    Mouse.release(MOUSE_MIDDLE);
  }
}

void MouseWheel() {
  n = digitalRead(ENCSIGA);
  if ((enc0ALast == LOW) && (n == HIGH)) {
      if (digitalRead(ENCSIGB) == LOW) {
          enc0Pos--;
      } else {
          enc0Pos++;
      }
      Mouse.move(0, 0, enc0Pos);
  }
  enc0Pos = 0;
  enc0ALast = n;
}

void CycleRGB() {
  static uint8_t hue;
  for(int i = 0; i < NUM_LEDS/2; i++) {
    // let's set an led value
    leds[i] = CHSV(hue++,255,255);

    // now, let's first 20 leds to the top 20 leds, 
    leds(NUM_LEDS/2,NUM_LEDS-1) = leds(NUM_LEDS/2 - 1 ,0);
    FastLED.delay(33);
  }
}
```

**Reply by earlephilhower** [reply]
I haven't dug into your code, but off the bat please don't do `stdio_init_XXXX` in setup.  The Serial object runs the USB and POSIX systems already.,

**Reply by RyzerPhenix** [reply]
Okay, thank you for the hint. that was just a leftover from a lot of tries to get the serial communication to work.

**Comment by RyzerPhenix** [comment]
For evaluation puroses I also ran just this:

```cpp
void setup(){
   Serial.begin(115200);
}

void loop(){
   Serial.println("Hello World");
}
```
